src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -133,91 +126,242 @@ public void Add([NotNull] Operation operation, [NotNull] object objectToApplyTo)
+                                var conversionResult = ConvertToActualType(genericTypeOfArray, value);
 
-                    // get value (it can be cast, we just checked that)
-                    var array = (IList)patchProperty.Property.ValueProvider.GetValue(patchProperty.Parent);
+                                if (!conversionResult.CanBeConverted)
+                                {
+                                    LogError(new JsonPatchError(
+                                        objectToApplyTo,
+                                        operationToReport,
+                                        Resources.FormatInvalidValueForProperty(value, path)));
+                                }
 
-                    if (appendList)
-                    {
-                        array.Add(conversionResult.ConvertedInstance);
-                    }
-                    else
-                    {
-                        // specified index must not be greater than the amount of items in the array
-                        if (positionAsInteger <= array.Count)
-                        {
-                            array.Insert(positionAsInteger, conversionResult.ConvertedInstance);
+                                // get value (it can be cast, we just checked that) 
+                                var array = treeAnalysisResult.Container.GetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent) as IList;
+
+                                if (appendList)
+                                {
+                                    array.Add(conversionResult.ConvertedInstance);
+                                    treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent, array);
+                                }
+                                else
+                                {
+                                    // specified index must not be greater than the amount of items in the
+                                    // array
+                                    if (positionAsInteger <= array.Count)
+                                    {
+                                        array.Insert(positionAsInteger, conversionResult.ConvertedInstance);
+                                        treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent, array);
+                                    }
+                                    else
+                                    {
+                                        LogError(new JsonPatchError(
+                                            objectToApplyTo,
+                                            operationToReport,
+                                            Resources.FormatInvalidIndexForArrayProperty(operationToReport.op, path)));
+                                    }
+                                }
+                            }
+                            else
```

Comment:
Some of this is hard to read because there's so much of a gap between the `if` and the `else`. It might benefit from being rewritten in the style of:

```
if (error_condition)
{
    <report error>
    return;
}

// Do more
```